### PR TITLE
Add allKeys method to RiakBucket

### PR DESCRIPTION
**Pull request by ExNexu**
http://docs.basho.com/riak/latest/dev/references/http/list-keys/

### CODE REVIEW

File: src/test/scala/com/scalapenos/riak/RiakBucketSpec.scala [hunk 1]
```diff
@@ -54,6 +54,28 @@ class RiakBucketSpec extends RiakClientSpecification with RandomKeySupport {
...
+      allKeys.keys(0) must beEqualTo(key)
+    }
+
+    "list all keys from an empty bucket" in {
+      val bucket = randomBucket
+
+      val allKeys = bucket.allKeys().await
+
+      allKeys.keys should be(Nil)
+    }
```
Comment (agemooij) [inline src/test/scala/com/scalapenos/riak/RiakBucketSpec.scala]: What about a test case for a bucket containing more than one properties? Something like four or five sounds fair...

File: src/main/scala/com/scalapenos/riak/internal/RiakHttpClientHelper.scala [hunk 2]
```diff
@@ -165,6 +165,25 @@ private[riak] class RiakHttpClientHelper(system: ActorSystem) extends RiakUriSup
...
+
+  def allKeys(server: RiakServerInfo, bucket: String): Future[RiakKeys] = {
+    httpRequest(Get(AllKeysUri(server, bucket))).map { response ⇒
+      response.status match {
+        case OK ⇒
+          response.entity.toOption.map { body ⇒
+            import spray.json._
+
+            body.asString.parseJson.convertTo[RiakKeys]
+          }.getOrElse(RiakKeys(Nil))
```
Comment (agemooij) [inline src/main/scala/com/scalapenos/riak/internal/RiakHttpClientHelper.scala]: Here you could use `response.entity.as[RiakKeys]` and drop the `spray.json` import, just like in `getBucketProperties` above.

File: src/main/scala/com/scalapenos/riak/RiakBucket.scala [hunk 3]
```diff
@@ -70,5 +70,9 @@ trait RiakBucket {
   def lastWriteWins_=(value: Boolean): Future[Unit] = setLastWriteWins(value)
   def setLastWriteWins(value: Boolean): Future[Unit] = setProperties(Set(LastWriteWins(value)))
 
-}
+  // ============================================================================
+  // Special operations (Use with care)
+  // ============================================================================
 
+  def allKeys(): Future[RiakKeys]
```
Comment (agemooij) [inline src/main/scala/com/scalapenos/riak/RiakBucket.scala]: What if a `RiakBucket` had an `unsafe` method that produced an instance of `UnsafeBucketOperations` that you could then call `allKeys()` on? Would that be too protective?

File: src/test/scala/com/scalapenos/riak/RiakBucketSpec.scala (hunk 1 above)
Comment (ExNexu) [inline src/test/scala/com/scalapenos/riak/RiakBucketSpec.scala]: Done!

File: src/main/scala/com/scalapenos/riak/internal/RiakHttpClientHelper.scala (hunk 2 above)
Comment (ExNexu) [inline src/main/scala/com/scalapenos/riak/internal/RiakHttpClientHelper.scala]: Cool, thanks!

File: src/main/scala/com/scalapenos/riak/RiakBucket.scala (hunk 3 above)
Comment (ExNexu) [inline src/main/scala/com/scalapenos/riak/RiakBucket.scala]: No, I like the idea! Let me know what you think of the implementation. Right now `UnsafeBucketOperations` extends `RiakBucket`, because I think in practice it is more handy.

**Comment by ExNexu** [comment]
Updated.

I think the build is actually fine. The failure looks like a random resolve issue to me.

File: src/test/scala/com/scalapenos/riak/UnsafeBucketOperationsSpec.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      }
+
+      val allKeys = unsafeBucketOperations.allKeys().await
+
+      keys.map { key ⇒
+        unsafeBucketOperations.delete(key).await
+      }
+
+      allKeys.keys.size must beEqualTo(numberOfKeys)
+      allKeys.keys.toSet must beEqualTo(keys.toSet)
```
Comment (agemooij) [inline src/test/scala/com/scalapenos/riak/UnsafeBucketOperationsSpec.scala]: If you are already comparing the full set, the previous size check doesn't seem necessary.

File: src/test/scala/com/scalapenos/riak/TestUtils.scala [hunk 5]
```diff
@@ -39,3 +39,11 @@ trait RandomKeySupport {
 
   def randomKey = randomUUID().toString
 }
+
+trait RandomBucketSupport {
+  self: RiakClientSpecification with RandomKeySupport ⇒
+
+  import java.util.UUID._
```
Comment (agemooij) [inline src/test/scala/com/scalapenos/riak/TestUtils.scala]: Why is this import here? It is not used AFAICS.

**Comment by agemooij** [comment]
Some minor remarks left but it looks pretty good like this. 

I just restarted the Travis build.

File: src/test/scala/com/scalapenos/riak/UnsafeBucketOperationsSpec.scala (hunk 4 above)
Comment (ExNexu) [inline src/test/scala/com/scalapenos/riak/UnsafeBucketOperationsSpec.scala]: Good point!

File: src/test/scala/com/scalapenos/riak/TestUtils.scala (hunk 5 above)
Comment (ExNexu) [inline src/test/scala/com/scalapenos/riak/TestUtils.scala]: Yeah, sorry!

**Comment by ExNexu** [comment]
Just cleaned up following your comments. Thank you! :-)
